### PR TITLE
feat(integer): add unsigned overflowing mul

### PR DESCRIPTION
**Pull request by tmontaigu**

### CODE REVIEW

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 1]
```diff
@@ -2034,6 +2051,65 @@ where
...
+            );
+
+            let decrypted_res: u64 = cks.decrypt(&ct_res);
+            let decrypted_overflow = cks.decrypt_bool(&overflow_res);
+
+            let (expected_clear, expected_overflow) =
+                overflowing_sum_slice_under_modulus(&clears, modulus);
+
+            assert_eq!(decrypted_res, expected_clear,
+            "Invalid result for sum of ciphertext, expecter {expected_clear} got {decrypted_res}");
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:2106]: expecter -> expected

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 2]
```diff
@@ -2034,6 +2051,65 @@ where
...
+            assert_eq!(decrypted_overflow, expected_overflow,
+            "Invalid result for overflow flag of sum of ciphertext, expecter {decrypted_overflow} got {expected_overflow}");
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:2108]: expecter -> expected

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 3]
```diff
@@ -2034,6 +2051,65 @@ where
...
+    // message_modulus^vec_length
+    let modulus = cks.parameters().message_modulus().0.pow(NB_CTXT as u32) as u64;
+
+    for len in [1, 2, 15, 16, 17, 64, 65] {
+        for _ in 0..NB_TESTS_SMALLER {
+            let clears = (0..len)
+                .map(|_| rng.gen::<u64>() % modulus)
+                .collect::<Vec<_>>();
+
+            // encryption of integers
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:2073]: could we add a case with trivial encryptions ?

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 4]
```diff
@@ -2140,6 +2216,134 @@ where
...
+                   decrypted_overflowed,
+                   expected_overflowed,
+                   "Invalid overflow flag result for overflowing_mul, for ({clear_lhs} -{clear_rhs}) % {modulus} 
+                    expected overflow flag {expected_overflowed}, got {decrypted_overflowed}"
+                );
+            assert_eq!(result_overflowed.0.degree.get(), 1);
+            assert_eq!(result_overflowed.0.noise_level(), NoiseLevel::NOMINAL);
+        }
+    }
+
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:2311]: do you have a few hard coded test cases maybe ?

**Review by IceTDrinker — COMMENTED** [review]
Looks generally good, some typos

File: tfhe/src/integer/server_key/radix_parallel/add.rs [hunk 5]
```diff
@@ -999,6 +974,296 @@ impl ServerKey {
...
+                    } else {
+                        self.key.create_trivial(0)
+                    };
+                    carry_ct.blocks.rotate_right(1);
+
+                    (s.clone(), carry_ct, out_carry)
+                })
+                .collect_into_vec(&mut tmp_out);
+
+            // tmp_out elements are tuple of 2 elements (message, carry)
```
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/add.rs:1130]: Yes the comment is outdated, tmp_out contains tupe of 3 elements (message , carries, last_block_carry) (s.clone() is the message extracted in place)

File: tfhe/src/integer/server_key/radix_parallel/add.rs [hunk 6]
```diff
@@ -563,6 +566,7 @@ impl ServerKey {
             self.compute_prefix_sum_hillis_steele(generates_or_propagates, sum_function);
         let mut last_block_out_carry = self.key.create_trivial(0);
         std::mem::swap(&mut carries_out[num_blocks - 1], &mut last_block_out_carry);
+        last_block_out_carry.degree = Degree::new(1);
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/add.rs:569]: thanks

File: tfhe/src/integer/server_key/radix_parallel/add.rs [hunk 7]
```diff
@@ -999,6 +974,302 @@ impl ServerKey {
...
+                                .for_each(|block| {
+                                    self.key.message_extract_assign(block);
+                                });
+                        },
+                        || {
+                            // Contrary to non overflowing version we always extract all carries
+                            // as we need to track overflows
+                            carry_ct.blocks[addition_range.clone()]
+                                // .par_iter_mut()
+                                .iter_mut()
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/add.rs:1108]: par iter is commented out ?

File: tfhe/src/integer/server_key/radix_parallel/add.rs [hunk 8]
```diff
@@ -999,6 +974,302 @@ impl ServerKey {
...
+                    } else {
+                        self.key.create_trivial(0)
+                    };
+                    carry_ct.blocks.rotate_right(1);
+
+                    (s.clone(), carry_ct, out_carry)
+                })
+                .collect_into_vec(&mut tmp_out);
+
+            // tmp_out elements are tuple of 4 elements (message, carry, last_block_carry)
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/add.rs:1136]: 4 -> 3 ?

File: tfhe/src/integer/server_key/radix_parallel/add.rs (hunk 7 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/add.rs:1108]: well spotted

**Review by IceTDrinker — APPROVED** [review]
Looks good thanks !

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

**Pull request merged**